src/main/java/seedu/toluist/model/Task.java

```diff
@@ -61,4 +61,22 @@ public void replaceTags(ArrayList<Tag> tags) {
     public boolean isOverdue() {
         return startDateTime != null && startDateTime.isBefore(LocalDateTime.now());
     }
+
+    public boolean isStringContainedInDescriptionIgnoreCase(String comparison) {
+        return this.description.toLowerCase().contains(comparison.toLowerCase());
+    }
+
+    public boolean isStringContainedInAnyTagIgnoreCase(String comparison) {
+        for (Tag tag : allTags) {
+            if (tag.tagName.toLowerCase().contains(comparison.toLowerCase())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public int compareTo(Task comparison) {
```

Comment:
A high level view of an algorithm where you dont need to consider duplicates is as follows:
For each task:
      if intersection of search terms with lis of words in description is not empty OR intersection of tag list witht tasks' tag list is nott empty:
                add task to foundTasks
Clearly, task is only added once, so you dont need to worry abt duplicates. An additional benefit is that tthe check for tag can be short circuited if the description alr matched